Refactor workspace settings

mollfpr opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
Refactoring workspace settings form to use `Form.js` and dismiss keyboard after form submission.

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/11195
$ https://github.com/Expensify/App/issues/11195#issuecomment-1263410831


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- UI looks as it did before the refactor
- Values can be added and edited
- Errors are highlighted correctly (input border)
- Error messages show up correctly
- Form alerts are displayed correctly
- Clicking the fix the errors link focuses the first input with error
- No duplicate submission of the form occurs (when it's already submitting)

**Saving a workspace name should close the keyboard**
1. Go to the workspace section
2. Click on the workspace name
3. Form input will come to address the change in the name
4. Once the desired name is input, press the save button
5. On the native app and mWeb verified that the keyboard is closed after pressing the save button

- [x] Verify that no errors appear in the JS console

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- UI looks as it did before the refactor
- Values can be added and edited
- Errors are highlighted correctly (input border)
- Error messages show up correctly
- Form alerts are displayed correctly
- Clicking the fix the errors link focuses the first input with error
- No duplicate submission of the form occurs (when it's already submitting)

**Saving a workspace name should close the keyboard**
1. Go to the workspace section
2. Click on the workspace name
3. Form input will come to address the change in the name
4. Once the desired name is input, press the save button
5. On the native app and mWeb verified that the keyboard is closed after pressing the save button

- [x] Verify that no errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->


https://user-images.githubusercontent.com/25520267/194132341-6104c3c5-3daf-4b9f-bfd8-a0b407048a29.mov



#### Mobile Web - Chrome
<!-- Insert screenshots of your changes on the web platform (from chrome mobile browser)-->
https://user-images.githubusercontent.com/25520267/194132064-0f2b9e74-49ff-427e-bd38-fdec5bfc2c2a.mov

#### Mobile Web - Safari
<!-- Insert screenshots of your changes on the web platform (from safari mobile browser)-->


https://user-images.githubusercontent.com/25520267/194132325-02ffceb8-b3d8-46fd-9292-321f98224807.mov



#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->


https://user-images.githubusercontent.com/25520267/194132130-26b35d1a-3760-44b7-bb90-6b2e7133db4f.mov



#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->


https://user-images.githubusercontent.com/25520267/194132155-422f49cf-2691-45d8-b962-64ff050c94ce.mov




#### Android
<!-- Insert screenshots of your changes on the Android platform-->

https://user-images.githubusercontent.com/25520267/194132170-f68b201c-c128-4610-a29d-fe3d0bd881f0.mov

mollfpr committed:
fix error link pressed callback

mollfpr committed:
adding option to use scrollview

mollfpr committed:
refactoring workspace settings form

mollfpr committed:
Merge branch 'main' of https://github.com/mollfpr/App into refactor-workspace-settings

mollfpr commented on src/pages/workspace/WorkspacePageWithSections.js:
```diff
@@ -56,6 +56,9 @@ const propTypes = {
     }).isRequired,
 
+    /** Option to not use the default scroll view  */
+    shouldNotUseScrollView: PropTypes.bool,
```
> is form being nested inside a ScrollView causing a problem?

Yes. The `ScrollView` in `WorkspacePageWithSections` makes the submit button of `Form.js` not placed on the bottom.
<img width="412" alt="Screen Shot 2022-10-12 at 17 07 51" src="https://user-images.githubusercontent.com/25520267/195315069-47b8ffd2-b33e-41de-a4eb-ab9bbf39aa98.png">

mollfpr replied:
@rushatgabhane Yup, we won’t need that eventually.

luacmartins replied:
@rushatgabhane the only other solution I came up with was to not use `WorkspacePageWithSections`, but I don't think that's better.

mollfpr replied:
@rushatgabhane I guess we can use `shouldUseScrollView` instead but need to update the other pages that use `WorkspacePageWithSections`. Are you on board with this?

mollfpr replied:
@rushatgabhane I agree with your suggestion! I'll update the PR, thanks!

mollfpr committed:
Merge branch 'main' of https://github.com/mollfpr/App into refactor-workspace-settings

mollfpr committed:
replacing `shouldNotUseScrollView` with `shouldUseScrollView`

mollfpr committed:
Adding `shouldUseScrollView` to current workspace pages

mollfpr committed:
reverting Form.js changes

mollfpr commented:
@rushatgabhane I updated the PR and also reverted the `Form.js` changes because the issue is fixed [here](https://github.com/Expensify/App/pull/11847).

rushatgabhane reviewed (changes requested):
@mollfpr bug: clicking fix the error message doesn't focus the text input

https://user-images.githubusercontent.com/29673073/196797769-5e2292c7-72e4-4feb-9a0f-293b1bf17af8.mov

mollfpr committed:
Merge branch 'main' of https://github.com/mollfpr/App into refactor-workspace-settings

mollfpr commented:
@rushatgabhane I just pulled the latest main, it should be fixed the issue by clicking the error text to focus the input.

mollfpr commented on src/pages/workspace/WorkspaceSettingsPage.js:
```diff
@@ -58,21 +52,22 @@ class WorkspaceSettingsPage extends React.Component {
         }
-        const name = this.state.name.trim();
-        const outputCurrency = this.state.currency;
+        const name = values.name.trim();
```
Good catch! On the `submit` function we should not worry about that, it will go through `validate` function first.

mollfpr committed:
adding check values.name

mollfpr commented on src/pages/workspace/WorkspaceSettingsPage.js:
```diff
@@ -58,21 +52,22 @@ class WorkspaceSettingsPage extends React.Component {
         const errors = {};
-        if (!this.state.name.trim().length) {
-            errors.nameError = true;
+        if (!values.name.trim().length) {
```
I just add more conditions to the `name` validation. It will catch the `null` or `undefined` value.

mollfpr commented:
@rushatgabhane Added the original issue test and update the checklist with the latest template!

mollfpr committed:
Merge branch 'main' of https://github.com/mollfpr/App into refactor-workspace-settings

rushatgabhane reviewed (approved):
@Luke9389 LGTM! 🎉

luacmartins reviewed (approved):
LGTM

luacmartins commented:
Tests passed. Removing `Emergency` label.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3316889751) to staging by @luacmartins in version: 1.2.19-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

Luke9389 commented on src/pages/workspace/WorkspacePageWithSections.js:
```diff
@@ -103,16 +107,20 @@ class WorkspacePageWithSections extends React.Component {
-
-                    </View>
-                </ScrollView>
+                {this.props.shouldUseScrollView
```
I think `this.props.shouldUseScrollView` could use a better name. 
It should indicate _why_ we want to use scroll view.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3332929926) to production by @chiragsalian in version: 1.2.19-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
